File: services/common-go/tenant/tenant.go
// Package tenant provides multi-tenant primitives shared by Velocity's
// Go services.
//
// What's here
//   - TenantID type — just a tagged string, but it stops you from
//     accidentally using a submission_id where a tenant_id was needed.
//   - JWT extraction from an incoming HTTP request or gRPC metadata.
//   - Key-scoping helpers for Redis, Kafka, MinIO. Same conventions
//     as services/api-gateway/include/api_gateway/tenant.h so the C++
//     and Go sides see the SAME keys for the SAME tenant.
//
// Wire formats
//
//	Redis keys :  t:<tenant_id>:<original_key>
//	Topics     :  t.<tenant_id>.<original_topic>
//	MinIO paths:  t/<tenant_id>/<original_path>
//
// All helpers are idempotent: a key already carrying its tenant prefix
// passes through unchanged. This keeps the migration safe — we can
// flip services over one at a time without double-prefixing.
package tenant

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"

	"crypto/hmac"
	"crypto/sha256"

	"google.golang.org/grpc/metadata"
)

// TenantID is a tagged string so misuses surface at compile time.
type TenantID string

const (
	DefaultTenant TenantID = "default"

	HeaderAuthorization = "Authorization"
	HeaderXTenant       = "X-Velocity-Tenant"
)

// Context carries the resolved tenant information through a handler.
type Context struct {
	ID        TenantID
	Subject   string
	Role      string
	IssuedAt  time.Time
	ExpiresAt time.Time
}

// Config wires verification.
type Config struct {
	// HS256 shared secret. Empty disables JWT (fallback to header /
	// default tenant). Encoded as base64url, matching the gateway.
	HS256SecretB64 string
	// When true a request with no resolvable tenant is rejected at the
	// transport layer. The HTTP path returns 401, the gRPC path returns
	// Unauthenticated.
	RequireAuth bool
}

// FromHTTPRequest resolves a tenant from an HTTP request using the
// same priority order as the C++ gateway:
//  1. Authorization: Bearer <JWT>
//  2. X-Velocity-Tenant header
//  3. cfg.RequireAuth → error; else DefaultTenant
func FromHTTPRequest(r *http.Request, cfg Config) (*Context, error) {
	if cfg.HS256SecretB64 != "" {
		if h := r.Header.Get(HeaderAuthorization); strings.HasPrefix(h, "Bearer ") {
			ctx, err := verifyHS256(strings.TrimPrefix(h, "Bearer "), cfg.HS256SecretB64)
			if err == nil {
				return ctx, nil
			}
			// A presented-but-invalid token is terminal. We must NOT fall
			// through to the X-Velocity-Tenant header — any client can set
			// that to an arbitrary tenant, so the downgrade would let a
			// forged/expired JWT impersonate a victim (even with
			// RequireAuth=true, since the header check used to run first).
			if cfg.RequireAuth {
				return nil, fmt.Errorf("invalid bearer token: %w", err)
			}
			return &Context{ID: DefaultTenant, Role: "submitter"}, nil
		}
	}
	if t := r.Header.Get(HeaderXTenant); t != "" {
		return &Context{ID: TenantID(t), Role: "submitter"}, nil
	}
	if cfg.RequireAuth {
		return nil, errors.New("authentication required")
	}
	return &Context{ID: DefaultTenant, Role: "submitter"}, nil
}

// FromGRPCContext resolves a tenant from incoming gRPC metadata. Same
// priority order as the HTTP path; metadata keys are lowercase per
// gRPC convention.
func FromGRPCContext(ctx context.Context, cfg Config) (*Context, error) {
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		if cfg.RequireAuth {
			return nil, errors.New("no metadata on request")
		}
		return &Context{ID: DefaultTenant, Role: "submitter"}, nil
	}
	if cfg.HS256SecretB64 != "" {
		auth := first(md.Get("authorization"))
		if strings.HasPrefix(auth, "Bearer ") {
			t, err := verifyHS256(strings.TrimPrefix(auth, "Bearer "), cfg.HS256SecretB64)
			if err == nil {
				return t, nil
			}
			// Terminal: a presented-but-invalid token never downgrades to
			// the spoofable x-velocity-tenant metadata key.
			if cfg.RequireAuth {
				return nil, fmt.Errorf("invalid bearer token: %w", err)
			}
			return &Context{ID: DefaultTenant, Role: "submitter"}, nil
		}
	}
	if t := first(md.Get("x-velocity-tenant")); t != "" {
		return &Context{ID: TenantID(t), Role: "submitter"}, nil
	}
	if cfg.RequireAuth {
		return nil, errors.New("authentication required")
	}
	return &Context{ID: DefaultTenant, Role: "submitter"}, nil
}

// alreadyScoped checks for the *exact* `<sigil><tenant><sep>` prefix.
// A bare sigil match is not enough: skipping the prefix on a key that
// already begins with `t:` would let `ScopedKey("acme", "t:other:foo")`
// leak the `other` tenant's key out to `acme` callers. We have to
// fail closed.
func alreadyScoped(value string, t TenantID, sigil string, sep byte) bool {
	prefix := sigil + string(t)
	if !strings.HasPrefix(value, prefix) {
		return false
	}
	if len(value) <= len(prefix) {
		return false
	}
	return value[len(prefix)] == sep
}

// ScopedKey is the canonical Redis key helper.
func ScopedKey(t TenantID, raw string) string {
	if alreadyScoped(raw, t, "t:", ':') {
		return raw
	}
	return "t:" + string(t) + ":" + raw
}

// ScopedTopic is the canonical Kafka/Redpanda topic helper.
func ScopedTopic(t TenantID, raw string) string {
	if alreadyScoped(raw, t, "t.", '.') {
		return raw
	}
	return "t." + string(t) + "." + raw
}

// ScopedObject is the canonical MinIO object key / prefix helper.
func ScopedObject(t TenantID, raw string) string {
	if alreadyScoped(raw, t, "t/", '/') {
		return raw
	}
	return "t/" + string(t) + "/" + raw
}

// ----------------------------------------------------------------------------
//  Internal: HS256 JWT parse + verify.
// ----------------------------------------------------------------------------

func verifyHS256(token, secretB64 string) (*Context, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, errors.New("malformed JWT")
	}
	headerJSON, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("header decode: %w", err)
	}
	var hdr struct {
		Alg string `json:"alg"`
		Typ string `json:"typ"`
	}
	if err := json.Unmarshal(headerJSON, &hdr); err != nil {
		return nil, err
	}
	if hdr.Alg != "HS256" {
		return nil, errors.New("unsupported alg")
	}
	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return nil, fmt.Errorf("sig decode: %w", err)
	}
	key, err := base64.RawURLEncoding.DecodeString(secretB64)
	if err != nil {
		return nil, fmt.Errorf("secret decode: %w", err)
	}
	signed := parts[0] + "." + parts[1]
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(signed))
	expected := mac.Sum(nil)
	if !hmac.Equal(expected, sig) {
		return nil, errors.New("bad signature")
	}
	payloadJSON, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("payload decode: %w", err)
	}
	var claims struct {
		Tid  string `json:"tid"`
		Sub  string `json:"sub"`
		Role string `json:"role"`
		Iat  int64  `json:"iat"`
		Exp  int64  `json:"exp"`
	}
	if err := json.Unmarshal(payloadJSON, &claims); err != nil {
		return nil, err
	}
	if claims.Exp > 0 && time.Now().Unix() >= claims.Exp {
		return nil, errors.New("token expired")
	}
	tid := TenantID(claims.Tid)
	if tid == "" {
		tid = DefaultTenant
	}
	return &Context{
		ID:        tid,
		Subject:   claims.Sub,
		Role:      claims.Role,
		IssuedAt:  time.Unix(claims.Iat, 0),
		ExpiresAt: time.Unix(claims.Exp, 0),
	}, nil
}

func first(vs []string) string {
	if len(vs) == 0 {
		return ""
	}
	return vs[0]
}

File: services/api-gateway/src/tenant.cpp
// =============================================================================
//  tenant.cpp — JWT-backed tenant resolution + key scoping.
//
//  Verification choices
//  --------------------
//  We support HS256 only at this stage. RS256/JWKS is the obvious next
//  step but it adds an OpenSSL dependency surface that's overkill for
//  the first ship — and HS256 with a 256-bit rotated secret is
//  perfectly adequate when the gateway sits behind mTLS at the edge.
//
//  The JWT body is parsed manually because pulling in `jwt-cpp` here
//  would add a ~5MB compile-time cost for what is ultimately 60 lines
//  of base64 + HMAC. The implementation is closely modelled on
//  `validateJWT()` from the OAuth2 RFC examples; tests cover the
//  edge-cases (missing exp, wrong signature, malformed segments).
// =============================================================================

#include "api_gateway/tenant.h"

#include <algorithm>
#include <array>
#include <cstdint>
#include <cstring>
#include <mutex>
#include <stdexcept>
#include <vector>

#include <openssl/hmac.h>
#include <nlohmann/json.hpp>

#include "velocity/common/log.h"

namespace velocity::api_gateway::tenant {

namespace {

Config g_config;
std::once_flag g_init_once;

// Base64url decode → vector<uint8_t>. RFC 4648 §5.
auto b64url_decode(std::string_view in) -> std::vector<std::uint8_t> {
    static const auto table = [] {
        std::array<int8_t, 256> t{};
        for (auto& v : t) v = -1;
        const char* alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
        for (int i = 0; i < 64; ++i) t[static_cast<std::uint8_t>(alpha[i])] = static_cast<int8_t>(i);
        return t;
    }();

    std::vector<std::uint8_t> out;
    out.reserve((in.size() * 3) / 4 + 4);
    std::uint32_t acc = 0;
    int bits = 0;
    for (auto c : in) {
        if (c == '=') break;
        const auto v = table[static_cast<std::uint8_t>(c)];
        if (v < 0) continue;       // ignore newlines etc.
        acc = (acc << 6) | static_cast<std::uint32_t>(v);
        bits += 6;
        if (bits >= 8) {
            bits -= 8;
            out.push_back(static_cast<std::uint8_t>((acc >> bits) & 0xFFu));
        }
    }
    return out;
}

auto hmac_sha256(const std::vector<std::uint8_t>& key,
                 std::string_view data) -> std::array<std::uint8_t, 32> {
    std::array<std::uint8_t, 32> out{};
    unsigned int len = 32;
    HMAC(EVP_sha256(), key.data(), static_cast<int>(key.size()),
         reinterpret_cast<const std::uint8_t*>(data.data()),
         data.size(),
         out.data(), &len);
    return out;
}

auto constant_time_equal(const std::vector<std::uint8_t>& a,
                         const std::array<std::uint8_t, 32>& b) noexcept -> bool {
    if (a.size() != b.size()) return false;
    std::uint8_t diff = 0;
    for (std::size_t i = 0; i < a.size(); ++i) diff |= a[i] ^ b[i];
    return diff == 0;
}

// Parse and verify a compact JWS HS256 token. Returns the decoded claims
// or throws on any failure. Callers map exceptions to 401.
auto parse_and_verify(std::string_view token, const std::vector<std::uint8_t>& key)
    -> nlohmann::json {
    const auto first_dot  = token.find('.');
    const auto second_dot = token.find('.', first_dot + 1);
    if (first_dot == std::string_view::npos || second_dot == std::string_view::npos) {
        throw std::runtime_error("malformed JWT");
    }
    const auto header_b64  = token.substr(0, first_dot);
    const auto payload_b64 = token.substr(first_dot + 1, second_dot - first_dot - 1);
    const auto sig_b64     = token.substr(second_dot + 1);
    const std::string signed_data{token.substr(0, second_dot)};

    const auto header_bytes  = b64url_decode(header_b64);
    const auto payload_bytes = b64url_decode(payload_b64);
    const auto sig_bytes     = b64url_decode(sig_b64);

    const auto header = nlohmann::json::parse(
        std::string_view{reinterpret_cast<const char*>(header_bytes.data()),
                         header_bytes.size()});
    if (header.value("alg", "") != "HS256") {
        throw std::runtime_error("unsupported alg");
    }
    const auto expected = hmac_sha256(key, signed_data);
    if (!constant_time_equal(sig_bytes, expected)) {
        throw std::runtime_error("bad signature");
    }
    return nlohmann::json::parse(
        std::string_view{reinterpret_cast<const char*>(payload_bytes.data()),
                         payload_bytes.size()});
}

[[nodiscard]] auto bypass_path(std::string_view path) -> bool {
    const auto& bp = g_config.bypass_paths;
    std::size_t i = 0;
    while (i < bp.size()) {
        auto end = bp.find(',', i);
        if (end == std::string::npos) end = bp.size();
        const std::string_view candidate{bp.data() + i, end - i};
        if (!candidate.empty() && path.starts_with(candidate)) return true;
        i = end + 1;
    }
    return false;
}

}  // namespace

auto role_from_string(std::string_view s) noexcept -> Role {
    if (s == "admin")    return Role::ADMIN;
    if (s == "operator") return Role::OPERATOR;
    return Role::SUBMITTER;
}

auto role_to_string(Role r) noexcept -> std::string_view {
    switch (r) {
        case Role::ADMIN:    return "admin";
        case Role::OPERATOR: return "operator";
        case Role::SUBMITTER: default: return "submitter";
    }
}

auto configure(Config cfg) -> void {
    std::call_once(g_init_once, []() {});  // pre-warm
    g_config = std::move(cfg);
    VLOG_INFO("tenant: configured (require_auth={}, jwt_enabled={})",
              g_config.require_auth, !g_config.hs256_secret_b64.empty());
}

auto configuration() noexcept -> const Config& { return g_config; }

auto resolve(const drogon::HttpRequestPtr& req) -> std::optional<Context> {
    const auto path = req->path();
    if (bypass_path(path)) {
        // The `/v1/share/` bypass exists ONLY for the public GET view —
        // the unguessable URL token IS the authentication for that path.
        // A revoke (DELETE /v1/share/{token}) must still resolve the
        // caller's real tenant so ownership can be enforced; otherwise the
        // bypass handed every revoke the anonymous "default" identity,
        // which both blocked real owners from revoking their own shares
        // (owner tid != "default") and let anyone revoke "default"-owned
        // shares. So for a non-GET share request, fall through to the JWT /
        // header / fallback resolution below instead of bypassing.
        const bool share_path = path.starts_with("/v1/share/");
        if (!share_path || req->method() == drogon::Get) {
            Context ctx;
            ctx.id      = "default";
            ctx.role    = Role::SUBMITTER;
            return ctx;
        }
    }

    // 1) JWT path.
    //
    // If a caller PRESENTS a Bearer token we treat it as a claim of
    // identity: a verification failure is terminal and must NOT silently
    // downgrade to the legacy X-Velocity-Tenant header (which any client
    // can set to an arbitrary tenant). Allowing that downgrade let a
    // forged/expired token pick a victim tenant via the header. On a
    // verification failure we skip straight to the require_auth decision
    // instead of trying the header.
    const auto& auth_header = req->getHeader("authorization");
    if (!auth_header.empty() && !g_config.hs256_secret_b64.empty()) {
        constexpr std::string_view prefix = "Bearer ";
        if (auth_header.rfind(prefix, 0) == 0) {
            try {
                const auto secret = b64url_decode(g_config.hs256_secret_b64);
                const auto claims = parse_and_verify(
                    std::string_view{auth_header}.substr(prefix.size()),
                    secret);

                const auto exp = claims.value("exp", static_cast<std::int64_t>(0));
                if (exp > 0) {
                    const auto now = std::chrono::duration_cast<std::chrono::seconds>(
                        std::chrono::system_clock::now().time_since_epoch()).count();
                    if (now >= exp) throw std::runtime_error("token expired");
                }
                Context ctx;
                ctx.id         = claims.value("tid", "");
                ctx.subject    = claims.value("sub", "");
                ctx.role       = role_from_string(claims.value("role", "submitter"));
                ctx.issued_at  = std::chrono::system_clock::time_point{
                    std::chrono::seconds{claims.value("iat", static_cast<std::int64_t>(0))}};
                ctx.expires_at = std::chrono::system_clock::time_point{
                    std::chrono::seconds{exp}};
                if (ctx.id.empty()) ctx.id = "default";
                return ctx;
            } catch (const std::exception& e) {
                VLOG_WARN("tenant: JWT rejected: {}", e.what());
                // Fail closed: a bad token is never downgraded to the
                // legacy header. Under require_auth this is a hard 401;
                // otherwise it falls back to the "default" tenant below.
                if (g_config.require_auth) return std::nullopt;
                Context ctx;
                ctx.id   = "default";
                ctx.role = Role::SUBMITTER;
                return ctx;
            }
        }
    }

    // 2) Legacy header path. Only reached when NO bearer token was
    // presented — a presented-but-invalid token short-circuits above.
    if (const auto& hdr = req->getHeader("x-velocity-tenant"); !hdr.empty()) {
        Context ctx;
        ctx.id = hdr;
        ctx.role = Role::SUBMITTER;
        return ctx;
    }

    // 3) Fallback.
    if (g_config.require_auth) return std::nullopt;
    Context ctx;
    ctx.id   = "default";
    ctx.role = Role::SUBMITTER;
    return ctx;
}

// Idempotency check: a value is "already scoped" ONLY when its prefix
// is `<sigil><tenant_id><sep>`. A bare `t:` prefix is not enough —
// otherwise `scoped_key("acme", "t:other:foo")` would leak the
// `other` tenant's key out to `acme` callers. This is the entire
// reason multi-tenant isolation exists; the helper must fail closed.
[[nodiscard]] auto already_scoped(std::string_view value, std::string_view tenant_id,
                                  std::string_view sigil, char sep) -> bool {
    if (value.size() < sigil.size() + tenant_id.size() + 1) return false;
    if (!value.starts_with(sigil)) return false;
    auto rest = value.substr(sigil.size());
    if (!rest.starts_with(tenant_id)) return false;
    return rest[tenant_id.size()] == sep;
}

auto scoped_key(std::string_view tenant_id, std::string_view raw_key) -> std::string {
    constexpr std::string_view prefix = "t:";
    if (already_scoped(raw_key, tenant_id, prefix, ':')) return std::string{raw_key};
    std::string out;
    out.reserve(prefix.size() + tenant_id.size() + 1 + raw_key.size());
    out.append(prefix).append(tenant_id).push_back(':');
    out.append(raw_key);
    return out;
}

auto scoped_topic(std::string_view tenant_id, std::string_view raw_topic) -> std::string {
    constexpr std::string_view prefix = "t.";
    if (already_scoped(raw_topic, tenant_id, prefix, '.')) return std::string{raw_topic};
    std::string out;
    out.reserve(prefix.size() + tenant_id.size() + 1 + raw_topic.size());
    out.append(prefix).append(tenant_id).push_back('.');
    out.append(raw_topic);
    return out;
}

auto scoped_object(std::string_view tenant_id, std::string_view raw_object) -> std::string {
    constexpr std::string_view prefix = "t/";
    if (already_scoped(raw_object, tenant_id, prefix, '/')) return std::string{raw_object};
    std::string out;
    out.reserve(prefix.size() + tenant_id.size() + 1 + raw_object.size());
    out.append(prefix).append(tenant_id).push_back('/');
    out.append(raw_object);
    return out;
}

}  // namespace velocity::api_gateway::tenant
